contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+will look for matching federated LB services and update the CNAME record for the headless service with the
+relevant SRV records for all the LB services in the federation dns server.
+This list can certainly be updated to the first class API data, if the federation API objects and clients
+are updated and maintained separately to have additional fields/data then that present in k8s API.
+
+Once this happens, the stateful pods can discover peers by querying the governing service domain (federated
+headless service). Taking the same example listed earlier if the peer (or a user) queries the service
+```store.mynamespace.myfederation.svc.example.com```, the below SRV records will be returned.
+```
+store-c1-0.etcd.mynamespace.myfederation.svc.example.com
+store-c1-1.etcd.mynamespace.myfederation.svc.example.com
+store-c1-2.etcd.mynamespace.myfederation.svc.example.com
+store-c1-3.etcd.mynamespace.myfederation.svc.example.com
+store-c2-0.etcd.mynamespace.myfederation.svc.example.com
+store-c2-1.etcd.mynamespace.myfederation.svc.example.com
+store-c2-2.etcd.mynamespace.myfederation.svc.example.com
+store-c2-3.etcd.mynamespace.myfederation.svc.example.com
+store-c3-0.etcd.mynamespace.myfederation.svc.example.com
+store-c3-1.etcd.mynamespace.myfederation.svc.example.com
+store-c3-2.etcd.mynamespace.myfederation.svc.example.com
+```
+The global federation dns service will have SRV records and relevant A records listing IPs of the LBs.
+The local in-cluster kube-dns will have the SRV records and the relevant A records listing the local
+stateful set pod IPs. The local kube-dns will be able to respond with both the local and global records,
+when a query hits the local kube-dns. The peers can choose to connect to the local identities or the
+global ones as the need be.
+
+#### Handling additional LB services
+
+As listed in the above approach, the current option of exposing the statefulset pods across clusters is to
+assign a LB service to each replica pod.
+If the federated service controller is used (by creating the service in federation) to create these services,
+then as per the current behaviour the service will be created in each cluster.
+This leads to exactly one unused service per statefulset per cluster.
+There are two alternatives to circumvent this problem:
+
+1 - Cluster affinity is introduced for services such that the controller creates the service only
+in needed cluster. The need of cluster affinity and anti-affinity in general is discussed
+[elsewhere](https://github.com/kubernetes/kubernetes/issues/41442) also, and as of today is already
+available in federation.
+
+2 - Federated statefulset controller handles the creation of cluster local services, rather than the
+federated service controller. It also will need to handle the dns records for each LB service into the
+federation (or cloud provider specific) public dns server.
+
+We propose using alternative 1 listed here, as it fits broader scheme of things and is more consistent with
```

Comment:
I don't think these objects need to be treated any way different.
Say, a pod created by a deployment can be queried and listed like any other pod.
Same with the lb type automatically created by federated statefulset. 
Also, if a service with the identical name (but with other properties exists already), it will be overwritten, similar to the cascaded objects behaviour (cascading from federation into clusters or cascading from an object to another object)